README.md
[![Build Status](https://secure.travis-ci.org/apiaryio/pitboss.png)](http://travis-ci.org/apiaryio/pitboss)
[![Build Status](https://ci.appveyor.com/api/projects/status/nctklpxwtt14vv4r?svg=true)](https://ci.appveyor.com/project/Apiary/pitboss)

![Pitboss](http://s3.amazonaws.com/img.mdp.im/renobankclubinside4.jpg_%28705%C3%97453%29-20120923-100859.jpg)

# Pitboss-NG (next gen)

## A module for running untrusted code

```javascript
var Pitboss = require('pitboss-ng').Pitboss;

var untrustedCode = "var a = !true; a";

var sandbox = new Pitboss(untrustedCode, {
  memoryLimit: 32*1024, // 32 MB memory limit (default is 64 MB)
  timeout: 5*1000, // 5000 ms to perform tasks or die (default is 500 ms = 0.5 s)
  heartBeatTick: 100 // interval between memory-limit checks (default is 100 ms)
});

sandbox.run({
  context: {
    'foo': 'bar',
    'key': 'value' // context must be JSON.stringify positive
  },
  libraries: {
    myModule: path.join(__dirname, './my/own/module')
    // will be available as global "myModule" variable for the untrusted code
  }
}, function callback (err, result) {
  sandbox.kill(); // don't forget to kill the sandbox, if you don't need it anymore
});

// OR other option: libraries can be an array of system modules
sandbox.run({
  context: {},
  libraries: ['console', 'lodash'] // we will be using global "lodash" & "console"
}, function callback (err, result) {
  // finished, kill the sandboxed process
  sandbox.kill();
});
```

### Runs JS code and returns the last eval'd statement

```javascript
var assert = require('chai').assert;
var Pitboss = require('pitboss-ng').Pitboss;

var code = "num = num % 5;\nnum;"

var sandbox = new Pitboss(code);

sandbox.run({context: {'num': 23}}, function (err, result) {
  assert.equal(3, result);
  sandbox.kill(); // sandbox is not needed anymore, so kill the sandboxed process
});
```

### Allows you to pass you own libraries into sandboxed content

```javascript
var assert = require('chai').assert;
var Pitboss = require('pitboss-ng').Pitboss;

var code = "num = num % 5;\n console.log('from sandbox: ' + num);\n num;"

var sandbox = new Pitboss(code);

sandbox.run({context: {'num': 23}, libraries: ['console']}, function (err, result) {
  // will print "from sandbox: 5"
  assert.equal(3, result);
  sandbox.kill(); // sandbox is not needed anymore, so end it
});
```

### Handles processes that take too damn long

```javascript
var assert = require('chai').assert;
var Pitboss = require('pitboss-ng').Pitboss;

var code = "while(true) { num % 3 };";

var sandbox = new Pitboss(code, {timeout: 2000});
sandbox.run({context: {'num': 23}}, function (err, result) {
  assert.equal("Timedout", err);
  sandbox.kill();
});
```

### Doesn't choke under pressure (or shitty code)

```javascript
var assert = require('chai').assert;
var Pitboss = require('pitboss-ng').Pitboss;

var code = "Not a JavaScript at all!";

var sandbox = new Pitboss(code, {timeout: 2000});

sandbox.run({context: {num: 23}}, function (err, result) {
  assert.include(err, "VM Syntax Error");
  assert.include(err, "Unexpected identifier");
  sandbox.kill();
});
```

### Doesn't handle this! But 'ulimit' or 'pidusage' does!

```javascript
var assert = require('chai').assert;
var Pitboss = require('pitboss-ng').Pitboss;

var code = "var str = ''; while (true) { str = str + 'Memory is a finite resource!'; }";

var sandbox = new Pitboss(code, {timeout: 10000});

sandbox.run({context: {num: 23}}, function (err, result) {
  assert.equal("Process failed", err);
  sandbox.kill();
});
```

And since Pitboss-NG forks each process, ulimit kills only the runner


CHANGELOG.md
# 1.0.0

- BREAKING CHANGE: Supports NodeJS 8.10+
- BREAKING CHANGE: converted to ES6 Javascript, no coffee-script anymore
- changed dependency: from [csv](https://www.npmjs.org/package/csv) to [csv-parse](https://www.npmjs.org/package/csv-parse) 4.3.4
- upgraded dependency: [clone](https://www.npmjs.org/package/clone) to 2.1.2
- upgraded dependency: [pidusage](https://www.npmjs.org/package/pidusage) to 2.0.17
- upgraded devDependency: [chai](https://www.npmjs.org/package/chai) to 4.2.0
- upgraded devDependency: [mocha](https://www.npmjs.org/package/mocha) to 6.0.2
- removed devDependency: [coffee-script](https://www.npmjs.org/package/coffee-script)
- removed devDependency: [gulp](https://www.npmjs.org/package/gulp)
- removed devDependency: [gulp-coffee](https://www.npmjs.org/package/gulp-coffee)
- removed devDependency: [gulp-mocha](https://www.npmjs.org/package/gulp-mocha)
- removed devDependency: [gulp-util](https://www.npmjs.org/package/gulp-util)

# 0.3.2

- added direct support for memory management at win32/win64 platforms
- using [appVeyor](https://ci.appveyor.com/project/Apiary/pitboss) to test at Windows
- using [gulp](https://www.npmjs.org/package/gulp) for development and tasks, works cross-platform

# 0.3.1

- added npm-debug.log and *.log into gitignore/npmignore
- fixed typo in example code around `options.heartBeatTick`

# 0.3.0

- use [pidusage](https://www.npmjs.org/package/pidusage) to monitor memory footprint (supports *nix, Darwin, Win platform)
- introduce `pitboss.kill()` to be run after `pitboss.run` to cleanly kill sandboxed fork
- updated [README](README.md) with examples how to pass libraries into sandboxed code

# 0.2.1

- little changes to package.json regarding repository entry

# 0.2.0

- package renamed to `pitboss-ng`
- __works with NodeJS 0.10, 0.12, iojs__
- dropped support NodeJS 0.8
- devDependencies bumped
- using `clone` to really avoid any context sharing between fork and parent process
- `coffee-script` is no more production dependency, package is compiled before being published to NPM registry

# 0.1.5 (not released to NPM, only available at GitHub)

- use `coffee-script` from 1.6.3 to 1.8.0, updated postinstall task accordingly

# 0.1.4 (not released to NPM, only available at GitHub)

- guessing command to get sub-process information from `os.platform` = works on *nix/Darwin platforms now (still no Win support)

# 0.1.3 (not released to NPM, only available at GitHub)

- using `coffee-script` as dependency to compile `src/*.coffee` in NPM `postinstall` task

# 0.1.2 (not released to NPM, only available at GitHub)

- added option to pass `libraries` (required modules) to context of used VM (thus alowing to use various core/user-land modules)
- added `options.memoryLimit` (kBytes) to Pitboss, using child_process with `ps` (no Win/Darwin support)


test/pitboss_test.js
const assert = require('chai').assert;

const { Pitboss, Runner } = require('../lib/pitboss-ng');

describe('Pitboss running code', () => {
  let pitboss = null;
  const code = `// EchoTron: returns the 'data' variable in a VM
  if(typeof data === 'undefined') {
    var data = null
  };
  data`;

  before(function() {
    pitboss = new Pitboss(code);
  });

  after(function() {
    pitboss.kill();
  });

  it('should take a JSON encodable message', function(done) {
    pitboss.run(
      {
        context: {
          data: 'test',
        },
      },
      function(err, result) {
        if (err) {
          done(err);
          return;
        }
        pitboss.run(
          {
            context: {
              data: 456,
            },
          },
          function(err, resultB) {
            if (err) {
              done(err);
              return;
            }
            assert.strictEqual(result, 'test');
            assert.strictEqual(resultB, 456);
            done();
          }
        );
      }
    );
  });
});

describe('Pitboss trying to access variables out of context', function() {
  let pitboss = null;
  let myVar = null;

  before(function() {
    const code = `if (typeof myVar === 'undefined') {
        var myVar;
      };
      myVar = "fromVM";
      myVar`;
    myVar = 'untouchable';
    pitboss = new Pitboss(code);
  });

  after(function() {
    pitboss.kill();
  });

  it('should not allow for context variables changes', function(done) {
    pitboss.run(
      {
        context: {
          love: 'tender',
          myVar: myVar,
        },
      },
      function(err, result) {
        assert.equal(result, 'fromVM');
        assert.equal(myVar, 'untouchable');
        done();
      }
    );
  });
});

describe('Pitboss modules loading code', function() {
  let pitboss = null;
  const code = `os.platform();
  console.log(data);
  data`;

  beforeEach(function() {
    pitboss = new Pitboss(code);
  });

  afterEach(function() {
    pitboss.kill();
  });

  it('should not return an error when loaded module is used', function(done) {
    pitboss.run(
      {
        context: {
          data: 'test',
        },
        libraries: ['os'],
      },
      function(err, result) {
        assert.isNull(err);
        assert.equal(result, 'test');
        done();
      }
    );
  });

  it('should return an error when unknown module is used', function(done) {
    pitboss.run(
      {
        context: {
          data: 'test',
        },
        libraries: [],
      },
      function(err, result) {
        assert.isUndefined(result);
        assert.include(err.toString(), 'VM Runtime Error: ReferenceError:');
        assert.include(err.toString(), 'os is not defined');
        done();
      }
    );
  });
});

describe('Running dubius code', function() {
  let pitboss = null;
  const code = `// EchoTron: returns the 'data' variable in a VM
    if(typeof data === 'undefined') {
      var data = null
    };
    data`;

  before(function() {
    pitboss = new Pitboss(code);
  });

  after(function() {
    pitboss.kill();
  });

  it('should take a JSON encodable message', function(done) {
    pitboss.run(
      {
        context: {
          data: 123,
        },
        libraries: ['console'],
      },
      function(err, result) {
        assert.equal(result, 123);
        done();
      }
    );
  });
});

describe('Running shitty code', function() {
  let pitboss = null;
  const code = 'WTF< this in not even code;';

  before(function() {
    pitboss = new Pitboss(code);
  });

  after(function() {
    pitboss.kill();
  });

  it('should return the error', function(done) {
    pitboss.run(
      {
        context: {
          data: 123,
        },
      },
      function(err, result) {
        assert.include(err, 'VM Syntax Error: SyntaxError:');
        assert.include(err, 'Unexpected identifier');
        assert.isUndefined(result);
        done();
      }
    );
  });
});

describe('Running infinite loop code', function() {
  let runner = null;
  const code = `if (typeof infinite != 'undefined' && infinite === true) {
    var a = true, b;
    while (a) {
      b = Math.random() * 1000;
      "This is an never ending loop!"
    };
  }
  "OK"`;

  afterEach(function() {
    runner.kill(1);
    runner = null;
  });

  it('should timeout and restart fork', function(done) {
    runner = new Runner(code, {
      timeout: 1000,
    });

    runner.run(
      {
        context: {
          infinite: true,
        },
      },
      function(err, result) {
        assert.equal('Timedout', err);
        runner.run(
          {
            context: {
              infinite: false,
            },
          },
          function(err, result) {
            assert.equal(result, 'OK');
            done();
          }
        );
      }
    );
  });

  it('should happily allow for process failure (e.g. ulimit kills)', function(done) {
    runner = new Runner(code, {
      timeout: 1000,
    });

    runner.run(
      {
        context: {
          infinite: true,
        },
      },
      function(err, result) {
        assert.equal('Process Failed', err);
        runner.run(
          {
            context: {
              infinite: false,
            },
          },
          function(err, result) {
            assert.equal(result, 'OK');
            done();
          }
        );
      }
    );

    runner.proc.kill('SIGKILL');
  });
});

describe('Running code which causes memory leak', function() {
  let runner = null;

  before(function() {
    const code = `
      var a = 'a', b = true;
      while (b) {
        b = !!b;
        a = a + "--------------------------++++++++++++++++++++++++++++++++++a";
      };
      b`;
    runner = new Runner(code, {
      timeout: 15000,
      memoryLimit: 1024 * 100,
    });
  });

  after(function() {
    runner.kill(1);
  });

  it('should end with MemoryExceeded error', function(done) {
    runner.run(
      {
        context: {
          infinite: true,
        },
      },
      function(err, result) {
        assert.equal(err, 'MemoryExceeded');
        done();
      }
    );
  });
});


test/forkable_test.js
const { assert } = require('chai');
const { fork } = require('child_process');

describe('The forkable process', function() {
  const referenceErrorMsg = 'VM Runtime Error: ReferenceError:';
  const syntaxErrorMsg = 'VM Syntax Error: SyntaxError:';
  const typeErrorMsg = 'VM Runtime Error: TypeError:';

  beforeEach(function() {
    this.runner = fork('./lib/forkable.js');
  });

  afterEach(function() {
    if (this.runner != null) {
      if (typeof this.runner.kill === 'function') {
        this.runner.kill();
      }
    }
    this.runner = null;
  });

  describe('basic operation', function() {
    beforeEach(function() {
      this.code = `// EchoTron: returns the 'data' variable in a VM
      if(typeof data === "undefined") {
        var data = null
      };
      data`;
    });

    it('run without errors', function(done) {
      this.runner.on('message', function(msg) {
        assert.equal(msg.id, '123');
        assert.equal(msg.result, null);
        done();
      });
      this.runner.send({
        code: this.code,
      });
      this.runner.send({
        id: '123',
        context: {},
      });
    });
  });

  describe('running code that assumes priviledge', function() {
    beforeEach(function() {
      this.code = `require('http');
      123;`;
    });

    it('should fail on require', function(done) {
      this.runner.on('message', function(msg) {
        assert.equal(msg.id, '123');
        assert.equal(msg.result, null);
        assert.include(msg.error, 'require is not defined');
        assert.include(msg.error, referenceErrorMsg);
        done();
      });
      this.runner.send({
        code: this.code,
      });
      this.runner.send({
        id: '123',
        context: {},
      });
    });
  });

  describe('Running code that uses JSON global', function() {
    beforeEach(function() {
      this.code = 'JSON.stringify({});';
    });

    it('should work as expected', function(done) {
      this.runner.on('message', function(msg) {
        assert.equal(msg.id, '123');
        assert.equal(msg.result, '{}');
        done();
      });
      this.runner.send({
        code: this.code,
      });
      this.runner.send({
        id: '123',
        context: {},
      });
    });
  });

  describe('Running code that uses Buffer', function() {
    beforeEach(function() {
      this.code = `var buf = new Buffer();
      123;`;
    });

    it('should fail on require', function(done) {
      this.runner.on('message', function(msg) {
        assert.equal(msg.id, '123');
        assert.equal(msg.result, null);
        assert.include(msg.error, 'Buffer is not defined');
        assert.include(msg.error, referenceErrorMsg);
        done();
      });
      this.runner.send({
        code: this.code,
      });
      this.runner.send({
        id: '123',
        context: {},
      });
    });
  });
  describe('Running shitty code', function() {
    beforeEach(function() {
      this.code = "This isn't even Javascript!!!!";
    });

    it('should return errors on running of bad syntax code', function(done) {
      this.runner.on('message', function(msg) {
        assert.equal(msg.id, '123');
        assert.equal(msg.result, void 0);
        assert.include(msg.error, syntaxErrorMsg);
        assert.include(msg.error, 'Unexpected identifier');
        done();
      });
      this.runner.send({
        code: this.code,
      });
      this.runner.send({
        id: '123',
        context: {},
      });
    });
  });
  describe('Running runtime error code', function() {
    beforeEach(function() {
      this.code = `var foo = [];
      foo[data][123];`;
    });

    it('should happily suck up and relay the errors', function(done) {
      this.runner.on('message', function(msg) {
        assert.equal(msg.id, '123');
        assert.equal(msg.result, void 0);
        assert.include(msg.error, typeErrorMsg);
        assert.include(msg.error, "Cannot read property '123' of undefined");
        done();
      });
      this.runner.send({
        code: this.code,
      });
      this.runner.send({
        id: '123',
        context: {
          data: 'foo',
        },
      });
    });
  });

  describe('requiring libraries in context', function() {
    describe('from array', function() {
      beforeEach(function() {
        this.code = `if(vm == undefined){
          throw('vm is undefined');
        }
        null`;
      });

      it('should require and pass library to the context under variriable with module name', function(done) {
        this.runner.on('message', function(msg) {
          assert.equal(msg.id, '123');
          assert.equal(msg.result, null);
          assert.equal(msg.error, null);
          done();
        });
        this.runner.send({
          code: this.code,
        });
        this.runner.send({
          id: '123',
          context: {
            data: 'foo',
          },
          libraries: ['vm'],
        });
      });
    });

    describe('from object for specifiyng context variable name', function() {
      beforeEach(function() {
        this.code = `if(vmFooBar == undefined){
          throw('vmFooBar is undefined');
        }
        null`;
      });

      it('should require and pass library to the context under variable with key name', function(done) {
        this.runner.on('message', function(msg) {
          assert.equal(msg.id, '123');
          assert.equal(msg.result, null);
          assert.equal(msg.error, null);
          done();
        });
        this.runner.send({
          code: this.code,
        });
        this.runner.send({
          id: '123',
          context: {
            data: 'foo',
          },
          libraries: {
            vmFooBar: 'vm',
          },
        });
      });
    });

    describe('from unintentional other type', function() {
      beforeEach(function() {
        this.code = `var a = 'result';
        a`;
      });

      it('should raise and exception telling that it expects array or object', function(done) {
        this.runner.on('message', function(msg) {
          assert.equal(msg.id, '1234');
          assert.equal(msg.result, void 0);
          assert.equal(
            msg.error,
            'Pitboss error: Libraries must be defined by an array or by an object.'
          );
          done();
        });
        this.runner.send({
          code: this.code,
        });
        this.runner.send({
          id: '1234',
          context: {
            data: 'foo',
          },
          libraries: 'vm',
        });
      });
    });
  });
});


lib/forkable.js
const vm = require('vm');
const clone = require('clone');

let script = null;
let errorStatusMsg = null;
let errorStatus = 0;
let timeout = undefined;

const STATUS = { FATAL: 1 };

process.on('message', function onMessage(msg) {
  if (msg['code']) {
    if (msg['timeout']) {
      timeout = parseInt(msg['timeout'], 10);
    }
    create(msg['code']);
  } else {
    run(msg);
  }
});

function create(code) {
  var codeForVm, err;
  codeForVm = `"use strict";\n${code}`;
  try {
    if (vm.Script) {
      script = new vm.Script(codeForVm, {
        filename: 'sandbox',
        timeout
      });
    } else {
      script = vm.createScript(codeForVm);
    }
  } catch (err) {
    // Fatal, never try again
    errorStatus = STATUS['FATAL'];
    errorStatusMsg = `VM Syntax Error: ${err}`;
  }
};

function run(msg) {
  if (isFatalError()) {
    error(errorStatusMsg, msg.id);
    return false;
  }

  if (!script) {
    error("No code to run");
    return false;
  }

  if (msg.context == null) {
    msg.context = {};
  }

  const context = vm.createContext(clone(msg.context));

  if (msg.libraries) {
    if (Array.isArray(msg.libraries)) {
      msg.libraries.forEach((lib) => {        
        context[lib] = require(lib);
      });
    } else if (typeof msg.libraries === 'object') {
      Object.keys(msg.libraries).forEach((libKey) => {
        context[libKey] = require(msg.libraries[libKey]);
      });
    } else {
      return error("Pitboss error: Libraries must be defined by an array or by an object.", msg.id);
    }
  }
  try {
    const res = {
      result: script.runInNewContext(context || {}, {
        timeout: timeout
      }) || null, // script can return undefined, ensure it's null
      id: msg.id
    };
    message(res);
  } catch (err) {
    error(`VM Runtime Error: ${err}`, msg.id);
  }
};

function isFatalError() {
  if (errorStatus === STATUS['FATAL']) {
    return true;
  } else {
    return false;
  }
};

function error(msg, id = null) {
  return message({
    error: msg,
    id
  });
};

function message(msg) {
  process.send(msg);
};


examples/advanced.js
var Pitboss = require('../').Pitboss;

var untrustedCode = "if (typeof returnMe === 'undefined') {\n\
  var returnMe = { theArray: [] };\n\
}\n\
var sortAsNumbers = function (a, b) {\n\
  var aInt = parseInt(a, 10), bInt = parseInt(b, 10);\n\
  if (aInt === bInt) {\n\
    return 0;\n\
  } else if (aInt < bInt) {\n\
    return -1;\n\
  }\n\
  return 1;\n\
};\n\
console.log('message from VM: returnMe =', returnMe);\n\
\n\
returnMe.theArray.sort(sortAsNumbers);\n\
returnMe;\n\
";

var sandbox = new Pitboss(untrustedCode, {
  memoryLimit: 64*1024,
  timeout: 5*1000,
  heartBeatTick: 500
});

var returnMe = {
  theArray: ['-1', '-5', '10', '15', '-3', '0', '20', '21', '-7', '4']
};

sandbox.run({
  context: {
    returnMe: returnMe
  },
  libraries: ['console']
}, function callback (err, result) {
  sandbox.kill();
  if (err) {
    console.error('An Error in sandbox happened:', err);
    return;
  }
  console.log('Local variable is not sorted (touched):', returnMe.theArray);
  console.log('Result is an array of numbers (sorted):', result.theArray);
});


lib/pitboss-ng.js
const path = require('path');
const os = require('os');
const { fork, exec } = require('child_process');
const { EventEmitter } = require('events');

const clone = require('clone');
const pusage = require('pidusage');
const csvParse = require('csv-parse');

const isWin = ['win', 'win32', 'win64'].includes(os.platform());

exports.Pitboss = class Pitboss {
  constructor(code, options) {
    this.runner = new exports.Runner(code, options);
    this.queue = [];
    this.runner.on('completed', this.next.bind(this));
  }

  run({ context, libraries }, callback) {
    this.queue.push({
      context: context,
      libraries: libraries,
      callback: callback
    });
    this.next();
  }

  kill() {
    if (this.runner) {
      this.runner.kill(1);
    }
  }

  next() {
    if (this.runner.running) {
      return false;
    }
    const c = this.queue.shift();
    if (c) {
      this.runner.run({
        context: c.context,
        libraries: c.libraries
      }, c.callback);
    }
  }

};

// Can only run one at a time due to the blocking nature of VM
// Need to queue this up outside of the process since it's over an async channel
exports.Runner = class Runner extends EventEmitter {
  constructor(code, options = {}) {
    super();
    this.run = this.run.bind(this);
    this.disconnect = this.disconnect.bind(this);
    this.messageHandler = this.messageHandler.bind(this);

    // try to launch the subprocess again, BUT notify callback (if any)
    this.failedForkHandler = this.failedForkHandler.bind(this);

    this.timeout = this.timeout.bind(this);
    this.memoryExceeded = this.memoryExceeded.bind(this);
    this.notifyCompleted = this.notifyCompleted.bind(this);
    this.winMemory = this.winMemory.bind(this);

    this.code = code;

    this.options = clone(options);
    this.options.memoryLimit = options.memoryLimit != null ? options.memoryLimit : 64 * 1024; // memoryLimit is in kBytes, so 64 MB here
    this.options.timeout = options.timeout != null ? options.timeout : 500;
    this.options.heartBeatTick = options.heartBeatTick != null ? options.heartBeatTick : 100;

    this.callback = null;
    this.running = false;
    this.launchFork();
  }

  launchFork() {
    if (this.proc) {
      if (typeof this.proc.removeAllListeners === "function") {
        this.proc.removeAllListeners('exit');
        this.proc.removeAllListeners('message');
      }
      if (this.proc.connected) {
        this.proc.kill('SIGTERM');
      }
    }
    this.proc = fork(path.join(__dirname, '../lib/forkable.js'));

    this.proc.on('message', this.messageHandler);
    this.proc.on('exit', this.failedForkHandler);

    this.proc.send({
      code: this.code,
      timeout: this.options.timeout + 100
    });
  }

  run({ context, libraries }, callback) {
    if (this.running) {
      return false;
    }
    const id = Date.now().toString() + Math.floor(Math.random() * 1000);
    const msg = {
      context,
      libraries,
      id,
    };

    this.callback = callback || false;
    this.startTimer();
    this.running = id;
    this.proc.send(msg);
    return id;
  }

  disconnect() {
    if (this.proc && this.proc.connected) {
      this.proc.disconnect();
    }
  }

  kill(dieWithoutRestart) {
    if (this.proc && this.proc.connected) {
      if (dieWithoutRestart) {
        this.proc.removeAllListeners('exit');
        this.proc.removeAllListeners('message');
      }
      this.proc.kill("SIGTERM");
    }
    this.closeTimer();
  }

  messageHandler(msg) {
    this.running = false;
    this.closeTimer();
    this.emit('result', msg);
    if (this.callback) {
      const cb = this.callback;
      this.callback = false;
      if (msg.error) {
        cb(msg.error);
      } else {
        cb(null, msg.result);
      }
    }
    this.notifyCompleted();
  }

  failedForkHandler() {
    this.running = false;
    this.closeTimer(this.timer);
    this.launchFork();
    const error = this.currentError || "Process Failed";
    this.emit('failed', error);
    if (this.callback) {
      this.callback(error);
    }
    this.notifyCompleted();
  }

  timeout() {
    if (this.currentError == null) {
      this.currentError = "Timedout";
    }
    this.kill();
  }

  memoryExceeded() {
    if (!this.proc || this.proc.pid == null) {
      return;
    }

    const pid = this.proc.pid;

    if (isWin) {
      this.winMemory(pid, (err, stats = []) => {
        if (err) {
          if (this.running) { // still running and some error occurs
            console.error("Process memory usage command failed", err);
          }
        }
        if (!err && (stats && stats[0] && stats[0].memUsage > this.options.memoryLimit)) {
          this.currentError = "MemoryExceeded";
          this.kill();
        }
      });
      return;
    }

    pusage(pid, (err, stat = {}) => {
      if (err) {
        if (this.running) { // still running and some error occurs
          console.error("Process memory usage command failed", err);
        }
      }
      pusage.clear();
      // memoryLimit is in kBytes, whereas stat.memory in bytes
      if (!err && stat.memory > (this.options.memoryLimit * 1024)) {
        this.currentError = "MemoryExceeded";
        this.kill();
      }
    });
  }

  notifyCompleted() {
    this.running = false;
    this.emit('completed');
  }

  startTimer() {
    this.closeTimer();
    this.timer = setTimeout(this.timeout, this.options['timeout']);
    this.memoryTimer = setInterval(this.memoryExceeded, this.options['heartBeatTick']);
  }

  closeTimer() {
    if (this.timer) {
      clearTimeout(this.timer);
      this.timer = null;
    }
    if (this.memoryTimer) {
      clearInterval(this.memoryTimer);
      this.memoryTimer = null;
    }
  }

  winMemory(pid, cb) {
    const taskListPath = 'tasklist.exe ';

    const taskList = (arg, taskListCallback) => {
      exec(`${taskListPath}${arg}`, function taskListExecuted(err, stdout) {
        taskListCallback(err, stdout);
      });
    };

    const procStat = (procStatCallback) => {
      const arg = `/fi "PID eq ${pid}" /fo CSV`;
      const stats = [];
      taskList(arg, function taskListDone(err, stdout = '') {
        if (err || !stdout) {
          return;
        }
        csvParse(stdout, {
          skip_empty_lines: true
        }, function csvParseDone(err, rows = []) {
          if (err) {
            procStatCallback(err, stats);
            return;
          }
          if (rows.length > 0) {
            rows.forEach((row) => {
              let memVal;
              if (!(parseInt(row[1], 10) === pid)) {
                return;
              }
              if (row[4]) {
                memVal = `${row[4] || ''}`.toLowerCase().replace(',', '.').trim();
                if (memVal.includes('k')) {
                  memVal = 1000 * parseInt(memVal.slice(0, -1)); // because it was 1234.567 k
                } else if (memVal.includes('m')) {
                  memVal = 1000 * 1000 * parseInt(memVal.slice(0, -1), 10); // because it was 1234.567 M
                } else {
                  memVal = 1000 * parseFloat(memVal.slice(0, -1));
                }
              } else {
                // kiloBytes by default
                memVal = parseFloat(row[4]);
              }
              stats.push({
                name: row[0],
                pid: pid,
                memUsage: memVal
              });
            });
            procStatCallback(err, stats);
          } else {
            // fail silently
            procStatCallback(err, stats);
          }
        });
      });
    };
    procStat(cb);
  }

};
